src/plugins/english/divinedaolibrary.ts

```diff
@@ -1,125 +1,259 @@
+   * that successfully were fulfilled.
+   *
+   * @private
+   */
+  async asyncMap<T, U>(
+    collection: T[],
+    callbackfn: (value: T, index: number, array: T[]) => Promise<U>,
+  ): Promise<U[]> {
+    return (await Promise.allSettled(collection.map(callbackfn)))
+      .filter(
+        <U>(p: PromiseSettledResult<U>): p is PromiseFulfilledResult<U> =>
+          p.status === 'fulfilled',
+      )
+      .map(({ value }) => value);
+  }
 
-    if (searchTerm) {
-      novels = novels.filter(novel =>
-        novel.name.toLowerCase().includes(searchTerm.toLowerCase()),
-      );
+  /**
+   * DDL links to future (unpublished) chapters from its novel pages. To be
+   * able to report updates correctly, try to figure out which chapter was the
+   * actual latest one to be published.
+   *
+   * @private
+   * @returns the path value of the latest published chapter (or undefined)
+   */
+  async findLatestChapter(novelPath: string): Promise<string | undefined> {
+    const link = `${this.site}wp-json/wp/v2/categories?slug=${novelPath}`;
+    const guessCategory = await fetchApi(link).then(res => res.json());
+    if (guessCategory.length !== 1) {
+      return undefined;
     }
-    return novels;
+    const categoryId = guessCategory[0].id;
+    const chapterLink = `${this.site}wp-json/wp/v2/posts?categories=${categoryId}&per_page=1`;
+    const lastChapter = await fetchApi(chapterLink).then(res => res.json());
+    if (lastChapter.length !== 1) {
+      return undefined;
+    }
+    return lastChapter[0].slug;
   }
 
-  async popularNovels(): Promise<Plugin.NovelItem[]> {
-    const link = this.site + 'novels';
-
-    const body = await fetchApi(link).then(res => res.text());
+  /**
+   * Based on some basic information, grab all the available information about
+   * a novel. Mainly used with {@link asyncMap} to expand on novel links that
+   * were extracted from other places.
+   *
+   * @private
+   */
+  async grabNovel(baseNovel: {
+    name: string;
+    path: string;
+    getChapters?: boolean;
+  }): Promise<Plugin.SourceNovel> {
+    const link = `${this.site}wp-json/wp/v2/pages?slug=${baseNovel.path}`;
+    const data = await fetchApi(link).then(res => res.json());
+    if (data.length !== 1) {
+      return {
+        name: baseNovel.name,
+        path: baseNovel.path,
+      };
+    }
+    const content = parseHTML(data[0].content.rendered);
+    const excerpt = parseHTML(data[0].excerpt.rendered);
+    const image = content('img').first();
+    let chapters: Plugin.ChapterItem[] = [];
+    if (baseNovel.getChapters) {
+      const linkedChapters = content('li > span > a')
+        .map((_, anchorEl) => {
+          const path = this.getPath(anchorEl.attribs['href']);
+          if (!path) return;
+          return {
+            name: content(anchorEl).text(),
+            path,
+          } satisfies Plugin.ChapterItem;
+        })
+        .toArray();
```

Comment:
I think you juse need to filter out elements which doesnt have href for chapter.
Slicing chapter list with latest index might have trouble if the site have random releases (eg: .., 99, 100, 125)

Reply:
This already happens. When you return `undefined` (as is done on line 115) within a Cheerio `.map()` it will [not be part of the resulting array](https://github.com/cheeriojs/cheerio/blob/439fd196cc8e9dc4a4f02113e3275ee0c8fb95cb/src/api/traversing.ts#L684-L686). A Cheerio `.map()` is not like a standard JavaScript `.map()`.

Reply:
So all elements in linkedChapters should have a valid url and slicing this array might be unnecesssary?

Reply:
Oh, do you mean the slicing that happens a few lines below this comment? That is a completely separare thing.

The Divine Dao Library website lies about their chapters, as I wrote in the PR body. So what my code does is it parses all the chapter links, which might give links for chapters 1 through 200. Then as a second step in findLatestChapter we can figure out that actually chapter 186 is the last published one (as it is at the time of this comment for [Alchemy Master](https://www.divinedaolibrary.com/alchemy-master/), where any link after chapter 186 is on the page but shows a Not found).

So then the slice simply removes all the links after the one to 186.

Otherwise it would have reported 200 chapters for Alchemy Master and the application has no way to inform the reader about when chapters actually go live. Links would not change at all (even the titles are already in it, e.g. https://www.divinedaolibrary.com/alchemy-master-chapter-192-another-alchemy-duel/ which is a Not found at the time of writing).

Reply:
I think we are talking about different things. Are you talking about the slice that happens on line 124, just below the lines you left your comment on?

DDL lies about their chapters on novel pages. If you just parse the novel page for [Alchemy Master](https://www.divinedaolibrary.com/alchemy-master/) it will give you 200 **valid** links to 200 chapters. But the latest chapter actually published at time of writing is 186. All links 187 through 200 look like they are there, and even include titles in their URLs, but all result in a Not Found error when you visit them.

I am not slicing faullty links or anything like that. In `this.findLatestChapter` I try to detect the URL for the actually latest published chapter. And then `.slice()` to remove all the links after it on the list, because we are sure those will all just give errors.

This way LNReader can report a correct number of chapters and actually inform readers when new chapters come out. The old plugin was just permanently stuck at showing 200 chapters for Alchemy Master.

**In short:** all URLs are valid, but several URLs at the end of the list are placeholders and need to be cut off.

Reply:
got it